File: examples/async/kucoin_ws_private_readonly.py
"""Open a KuCoin private WebSocket user-data stream."""

import asyncio
import os

from dcex.ws import kucoin


async def main() -> None:
    """Open the private stream, subscribe to order updates, and print one message."""
    api_key = os.environ["KUCOIN_API_KEY"]
    api_secret = os.environ["KUCOIN_API_SECRET"]
    passphrase = os.environ["KUCOIN_API_PASSPHRASE"]

    async with kucoin.private(
        api_key=api_key,
        api_secret=api_secret,
        passphrase=passphrase,
    ) as ws:
        await ws.subscribe_orders()
        print(await ws.recv())


if __name__ == "__main__":
    asyncio.run(main())

File: crates/dcex/examples/kucoin_ws_private_readonly.rs
use std::env;
use std::time::Duration;

use dcex::ws::kucoin::KucoinPrivateWebSocket;

#[tokio::main]
async fn main() -> dcex::Result<()> {
    let api_key = env::var("KUCOIN_API_KEY")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;
    let api_secret = env::var("KUCOIN_API_SECRET")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;
    let passphrase = env::var("KUCOIN_API_PASSPHRASE")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;

    let mut ws =
        KucoinPrivateWebSocket::new(api_key, api_secret, passphrase, Duration::from_secs(10))?;
    ws.connect().await?;
    ws.subscribe_orders().await?;
    println!("{}", ws.recv().await?);
    ws.close().await?;
    Ok(())
}
